### PR TITLE
8337681: PNGImageWriter uses much more memory than necessary

### PR DESCRIPTION
**Pull request by gredler**
At some point the `JPEGImageWriter` was optimized to reduce duplication / copying of rasters and data buffers (see `JDK-6266748`). However, `PNGImageWriter` never received a similar optimization, and is making unnecessary copies of raster and data buffers whenever PNG images are written.

Measurements: In a local test, `PNGImageWriter.write( )` and callees were initially showing up as an allocation hotspot (~400k objects / ~65 MB allocated), but after these optimizations I'm seeing much better numbers (~400 objects / ~740 KB allocated).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8337681](https://bugs.openjdk.org/browse/JDK-8337681): PNGImageWriter uses much more memory than necessary (**Bug** - P4)

### Reviewers
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)
 * [Laurent Bourgès](https://openjdk.org/census#lbourges) (@bourgesl - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20432/head:pull/20432` \
`$ git checkout pull/20432`

Update a local copy of the PR: \
`$ git checkout pull/20432` \
`$ git pull https://git.openjdk.org/jdk.git pull/20432/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20432`

View PR using the GUI difftool: \
`$ git pr show -t 20432`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20432.diff">https://git.openjdk.org/jdk/pull/20432.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20432#issuecomment-2263784240)

### CODE REVIEW

File: test/jdk/javax/imageio/plugins/png/RasterReuseWriteTest.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+import javax.imageio.ImageWriteParam;
+import javax.imageio.ImageWriter;
+import javax.imageio.stream.ImageInputStream;
+import javax.imageio.stream.ImageOutputStream;
+import javax.imageio.stream.MemoryCacheImageOutputStream;
+
+public class RasterReuseWriteTest {
+
+    public static void main(String[] args) throws Exception {
+        test(BufferedImage.TYPE_INT_ARGB);
```
Comment (prrace) [inline test/jdk/javax/imageio/plugins/png/RasterReuseWriteTest.java:49]: can you also test INT_RGB at least ?

**Review by prrace — COMMENTED** [review]
I'm going to put it through our internal test system.

File: test/jdk/javax/imageio/plugins/png/RasterReuseWriteTest.java (hunk 1 above)
Comment (gredler) [reply test/jdk/javax/imageio/plugins/png/RasterReuseWriteTest.java:50]: No problem, I've added `TYPE_INT_RGB`, `TYPE_4BYTE_ABGR` and `TYPE_4BYTE_ABGR_PRE`.

**Review by prrace — APPROVED** [review]

**Comment by prrace** [comment]
I've approved but we do require 2 approvals for client fixes.

**Comment by gredler** [comment]
@prrace: Thanks!

@Team: Could another reviewer take a look at this PR and provide feedback?

**Comment by bourgesl** [comment]
I will have a look if nobody else ramps up...

**Review by bourgesl — APPROVED** [review]
Lgtm, just an overall look to png writer changes

**Comment by gredler** [comment]
@prrace I think this is ready to be merged. Could you have a look and sponsor the PR if everything still looks good to you? Thanks!

File: src/java.desktop/share/classes/com/sun/imageio/plugins/png/PNGImageWriter.java [hunk 2]
```diff
@@ -919,17 +920,28 @@ private void encodePass(ImageOutputStream os,
 
         int bitDepth = metadata.IHDR_bitDepth;
         for (int row = minY + yOffset; row < minY + height; row += ySkip) {
-            Rectangle rect = new Rectangle(minX, row, width, 1);
-            Raster ras = image.getData(rect);
+            Raster ras;
+            if (image instanceof BufferedImage bi) {
```
Comment (mrserb) [inline src/java.desktop/share/classes/com/sun/imageio/plugins/png/PNGImageWriter.java:924]: Do we really need to run these two blocks in the loop for each iteration?

Comment (gredler) [reply src/java.desktop/share/classes/com/sun/imageio/plugins/png/PNGImageWriter.java:924]: Hi Sergey! The current logic does make it very clear where and how `ras` is initialized, which is nice. The top two initialization blocks (for `BufferedImage`s and tiled images) could in theory be hoisted outside the loop, but the third one cannot (and we might have to jump through some hoops to guarantee to the compiler that the variable is initialized if we start initializing it sometimes in one place and sometimes in another).

AFAICT the `BufferedImage.getRaster()` call is not doing any work (just providing a reference to an instance variable), but the `getTile(...)` call might be doing some extra work, depending on the specific instance type (but in any event will not be copying raster data, which was by far the biggest waste here).

**Pull request closed without merging**